### DISCUSSION TITLE
Using Caliper with Cray compiler

### DISCUSSION THREAD

**Discussion by mastino** [LLNL/Caliper#347] [Q&A]
Hello,

I am trying to measure an OpenMP application with Caliper (from git on Nov 18 2020). The Caliper build and the benchmark are built with CCE 10.2. When the code runs with Caliper instrumentation it hangs and does not finish. Using the gcc compiler I have not had these problems. Do you have recommendations for debugging this issue?

Thanks!

**Comment by daboehme** [comment]
Hi @mastino , can you run the program with `CALI_LOG_VERBOSITY=2` set as an environment variable? That should give us debug output that might tell us where it hangs. Also, you can try running the program in a debugger and try to get a stack trace once it hangs.

**Reply by mastino** [reply]
Here is the caliper output:
`== CALIPER: Available services: alloc,aggregate,cpuinfo,memusage,papi,topdown,libpfm,debug,env,event,loop_monitor,region_monitor,textlog,recorder,report,timestamp,statistics,spot,trace,validator,measurement_template,mpi,mpireport,mpiflush
== CALIPER: Initialized
== CALIPER: Creating channel default
== CALIPER: default: snapshot scopes: process thread
== CALIPER: default: Registered trace service
== CALIPER: default: event: Marked attribute annotation
== CALIPER: default: event: Marked attribute function
== CALIPER: default: event: Marked attribute loop
== CALIPER: default: event: Marked attribute statement
== CALIPER: default: Registered event trigger service
== CALIPER: papi: Found 1 event codes for 1 PAPI component(s)
== CALIPER: papi: Creating eventset with 1 events for component 0 (perf_event)
== CALIPER: default: Registered papi service
== CALIPER: default: Registered report service`

The backtrace is below it seems to be getting stuck on my CALI_MARK_END() call

`Thread 1 "test_linear.out" received signal SIGINT, Interrupt.
0x0000400000479220 in cali::MetadataTree::MetadataTreeImpl::get_path(unsigned long, cali::Attribute const*, cali::Variant const*, cali::Node*) ()
    at /home/gravelle/soft/src/Caliper/src/caliper/MetadataTree.cpp:377
377	            for (node = parent->first_child(); node && !node->equals(attr[i].id(), data[i]); node = node->next_sibling())
(gdb) bt
#0  0x0000400000479220 in cali::MetadataTree::MetadataTreeImpl::get_path(unsigned long, cali::Attribute const*, cali::Variant const*, cali::Node*) ()
    at /home/gravelle/soft/src/Caliper/src/caliper/MetadataTree.cpp:377
#1  0x0000400000476a14 in cali::MetadataTree::get_path(unsigned long, cali::Attribute const*, cali::Variant const*, cali::Node*) ()
    at /home/gravelle/soft/src/Caliper/src/caliper/MetadataTree.cpp:599
#2  0x00004000004479e8 in cali::Caliper::make_record(unsigned long, cali::Attribute const*, cali::Variant const*, cali::SnapshotRecord&, cali::Node*) ()
    at /home/gravelle/soft/src/Caliper/src/caliper/Caliper.cpp:1330
#3  0x00004000003ca48c in (anonymous namespace)::EventTrigger::pre_end_cb(cali::Caliper*, cali::Channel*, cali::Attribute const&, cali::Variant const&) ()
    at /home/gravelle/soft/src/Caliper/src/services/event/EventTrigger.cpp:215
#4  0x00004000003cb22c in (anonymous namespace)::EventTrigger::event_trigger_register(cali::Caliper*, cali::Channel*)::{lambda(cali::Caliper*, cali::Channel*, cali::Attribute const&, cali::Variant const&)#3}::operator() () at /home/gravelle/soft/src/Caliper/src/services/event/EventTrigger.cpp:287
#5  0x00004000003c8e28 in std::_Function_handler<void (cali::Caliper*, cali::Channel*, cali::Attribute const&, cali::Variant const&), (anonymous namespace)::EventTrigger::event_trigger_register(cali::Caliper*, cali::Channel*)::{lambda(cali::Caliper*, cali::Channel*, cali::Attribute const&, cali::Variant const&)#3}>::_M_invoke(std::_Any_data const&, cali::Caliper*&&, cali::Channel*&&, cali::Attribute const&, cali::Variant const&) ()
    at /opt/gcc/8.1.0/snos/include/g++/bits/std_function.h:297
#6  0x00004000004450c8 in cali::Caliper::end(cali::Attribute const&) () at /home/gravelle/soft/src/Caliper/src/caliper/Caliper.cpp:1062
#7  0x00004000004832e8 in cali_end_region () at /home/gravelle/soft/src/Caliper/src/caliper/cali.cpp:385
#8  0x0000000000403308 in multiply_matrix_t_cray$mt$p0001 () at linear_algebra.c:101
#9  0x0000400000c98d3c in _cray$mt_start_one_code_parallel () from /opt/cray/pe/cce-sve/10.0.2/cce/aarch64/lib/libcraymp.so.1
#10 0x0000400000c9a880 in _cray$mt_execute_parallel_with_proc_bind () from /opt/cray/pe/cce-sve/10.0.2/cce/aarch64/lib/libcraymp.so.1
#11 0x0000000000402ed0 in multiply_matrix_t (mat_a=0x94dd00, rows_a=8192, cols_a=8192, mat_b=0x209ddd40, cols_b=8192, mat_c=0x40a6dd80)
    at linear_algebra.c:83
#12 0x0000000000402344 in test_multiply_2 () at test_linear_algebra.c:218
#13 0x0000000000401338 in main (argc=<optimized out>, argv=<optimized out>) at test_linear_algebra.c:70`

**Reply by mastino** [reply]
I can dump this to a text file if that would be easier

**Reply by daboehme** [reply]
Hm, OK. This ends up in a loop which can become a performance bottleneck if there are a lot of regions, but it's strange that it would hang. Also doesn't explain why it would work with gcc but not with CCE.

You can try setting `CALI_EVENT_ENABLE_SNAPSHOT_INFO=false` as a workaround and see if that helps.

Still, I'm not sure what the root cause is. Is this a code you can share?

**Reply by jrmadsen** [reply]
Just glancing at line ~377, my best guess would be that you have a duplicate siblings in the tree, e.g. ABA so that the `B->next_sibling()` returns to the first A. Maybe caused by a buffer overrun elsewhere where CCE offers less protection than GCC since all those proprietary compilers are basically just GCC/Clang with more safeguards off by default. You could probably verify by something like:

```
std::set<decltype(node)> _history{};
for (node = parent->first_child(); node && !node->equals(attr[i].id(), data[i]); )
{
   _history.insert(node);
   node = node->next_sibling();
   if(_history.count(node) > 0) throw std::runtime_error("Duplicate!");
}
```

**Reply by mastino** [reply]
Yes we can share code:
https://github.com/HPCL/benchmarks/blob/master/kernels/mm_blocks/linear_algebra.c#L63
This function is the only one we are measuring at the moment and it is only called once, so there really shouldn't be complex sets of regions. 

I tried your workaround @daboehme but it is seg faulting, so I will debug more in the morning.

**Reply by mastino** [reply]
I'm not sure what cause the segi fault last night, but the workaround appears to have worked and I haven't been able to replicate it this morning.

**Reply by mastino** [reply]
I was able to replicate the crash. At first I thought it was a memory leak on my side (this has been fixed), and the bt references: cali::Caliper::GlobalData::S_TLSObject::~S_TLSObject()

```
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc

Thread 1 "test_linear.out" received signal SIGABRT, Aborted.
0x00004000013f2dbc in raise () from /lib64/libc.so.6
(gdb) bt
#0  0x00004000013f2dbc in raise () from /lib64/libc.so.6
#1  0x00004000013e0928 in abort () from /lib64/libc.so.6
#2  0x00004000018326c0 in __gnu_cxx::__verbose_terminate_handler ()
    at ../../../../cray-gcc-cross-aarch64-8.1.0-201808020743.4afb0374114bb/libstdc++-v3/libsupc++/vterminate.cc:95
#3  0x000040000183083c in __cxxabiv1::__terminate (handler=<optimized out>)
    at ../../../../cray-gcc-cross-aarch64-8.1.0-201808020743.4afb0374114bb/libstdc++-v3/libsupc++/eh_terminate.cc:47
#4  0x0000400001830888 in std::terminate () at ../../../../cray-gcc-cross-aarch64-8.1.0-201808020743.4afb0374114bb/libstdc++-v3/libsupc++/eh_terminate.cc:57
#5  0x0000400001830894 in __cxxabiv1::__unexpected (handler=<optimized out>)
    at ../../../../cray-gcc-cross-aarch64-8.1.0-201808020743.4afb0374114bb/libstdc++-v3/libsupc++/eh_terminate.cc:63
#6  0x0000400001830774 in __cxxabiv1::__cxa_call_unexpected (exc_obj_in=<optimized out>)
    at ../../../../cray-gcc-cross-aarch64-8.1.0-201808020743.4afb0374114bb/libstdc++-v3/libsupc++/eh_personality.cc:759
#7  0x0000400000456304 in cali::Caliper::GlobalData::S_TLSObject::~S_TLSObject() () at /home/gravelle/soft/src/Caliper/src/caliper/Caliper.cpp:542
#8  0x00004000013f5a58 in __cxa_finalize () from /lib64/libc.so.6
#9  0x00004000002ba9f0 in __do_global_dtors_aux () from /projects/artab/soft/Caliper/a64-cce10.0.2-mvapich2.3.4/lib64/libcaliper.so.2
#10 0x000040000000e224 in _dl_fini () from /lib/ld-linux-aarch64.so.1
#11 0x00004000013f5450 in __run_exit_handlers () from /lib64/libc.so.6
#12 0x00004000013f55b4 in exit () from /lib64/libc.so.6
#13 0x00004000013e0d68 in __libc_start_main () from /lib64/libc.so.6
#14 0x0000000000401038 in _start ()
```

**Reply by mastino** [reply]
I've been trying to debug a bit more and found that caliper seems to think my markers are nested. The json looks a bit like this:

`{"cali.attribute.name":"thread_id/papi.PAPI_TOT_CYC/thread_id/papi.PAPI_TOT_CYC/thread_id","cali.attribute.prop":"85/85/85/85/85","cali.attribute.type":"double/uint/int/usr/string/uint/uint/uint/double/uint/int/usr/string/uint/uint/uint/usr/double/uint/uint/int/usr/string/string","class.aggregatable":"true/true","path":"block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm","papi.PAPI_TOT_CYC":0,"thread_id":5},
{"cali.attribute.name":"thread_id/papi.PAPI_TOT_CYC/thread_id/papi.PAPI_TOT_CYC/thread_id/papi.PAPI_TOT_CYC/thread_id/papi.PAPI_TOT_CYC/thread_id/papi.PAPI_TOT_CYC/thread_id/papi.PAPI_TOT_CYC","cali.attribute.prop":"85/85/85/85/85/85/85/85/85/85/85/85","cali.attribute.type":"double/uint/int/usr/string/uint/uint/uint/double/uint/int/usr/string/uint/uint/uint/double/uint/int/usr/string/uint/uint/uint/double/uint/int/usr/string/uint/uint/uint/double/uint/int/usr/string/uint/uint/uint/double/uint/int/usr/string/uint/uint/uint/double","class.aggregatable":"true/true/true/true/true/true","path":"block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm/block_mm"},`

**Reply by daboehme** [reply]
That's wild. Definitely very much not how this should look like. Thanks @mastino for debugging this. I suspect some weird race condition, broken atomics (we had an IBM compiler a while back where compare-and-swap was broken and caused havok) or somehow the OpenMP implementation messes with Caliper's thread-local storage. If even the attribute metadata gets messed up like that it could well be that we're creating some infinite loop like @jrmadsen suspected. 

Does this still happen with just 1 thread?

**Reply by mastino** [reply]
It does not seem to happen with just one thread

**Reply by jrmadsen** [reply]
Ah I think I know what's going on. Try explicitly initializing caliper before you enter the OMP region

**Reply by jrmadsen** [reply]
I think Caliper is lazily initializing on a child thread so it thinks that thread is the "main" thread and then when the threads get deleted, that "main" thread gets deleted before the other threads so the child threads now have pointers or something that no longer exists. @daboehme Do you still rely on `__attribute__((constructor))` to establish the main thread? Cray probably doesn't support that ctor whereas GNU (obviously) does.

**Reply by mastino** [reply]
@jrmadsen I have an attribute created before the omp region, is that not sufficient? Is there still a cali_init() function? I can't seem to find it in the documentation

**Reply by jrmadsen** [reply]
I have had a lot of issues in the past like this when the first instance of a timemory component happens on a child thread, especially when all the measurements only happen on child threads and especially with non-GNU and non-Clang compilers since the ctor attribute wasn't necessarily supported. I know Caliper has the same sort of issues with static lazy initialization (@daboehme maybe someday we'll figure out how to do it 100% properly) but I've resolved a ton of my issues once I replaced relying on the ctor attribute with something like:

```cpp
// void library_init_function() __attribute__((constructor));

void library_init_function();

namespace
{
// use comma operator to execute library_init_function() when
// the static initialization of _library_init_variable happens
static bool _library_init_variable = (library_init_function(), true);
}
```

**Reply by jrmadsen** [reply]
@mastino Yeah, try calling `cali_init()` although I would guess that the attribute creation would initialize caliper but @daboehme would obviously know better than me. I just know that @daboehme and I have struggled a lot with TLS initialization bc it is really, really difficult to robustly configure an implicit and lazy initialization system where data collection happen in thread-local memory (to avoid locking) and adds itself into parent thread. If `cali_init()` doesn't fix it, I'd suggest (as a debug exercise) seeing it creating a region on the main thread fixes it -- that would provide the evidence that this indeed another one of the TLS initialization issues.

**Reply by jrmadsen** [reply]
e.g.:

```cpp
CALI_MARK_BEGIN("dummy");
puts("dummy");
CALI_MARK_END("dummy");

#pragma omp parallel private(iii,jjj,ii,jj,i,j,k)
{
#ifdef USE_CALI_REG
CALI_MARK_BEGIN("block_mm");
#endif

  #pragma omp for
  for (int i = 0; i < rows_a; i++ ) {
    for (int jjj = 0; jjj < cols_b; jjj = jjj + BLOCK_ROWS) 
      for (int j = jjj; j < min(cols_b, jjj + BLOCK_ROWS); j++) 
        for (int kkk = 0; kkk < cols_a; kkk = kkk + BLOCK_COLS) {
          #pragma omp simd
          // #pragma omp simd aligned(mat_a, mat_b, mat_c: 64)
          for (int k = kkk; k < min(cols_a,kkk + BLOCK_COLS); k++)
            mat_c[i][j] = mat_c[i][j] + mat_a[i][k] * mat_b[j][k];
        }
  } // i

#ifdef USE_CALI_REG
CALI_MARK_END("block_mm");
}
```

**Reply by mastino** [reply]
Thanks for the help; I tried with cali_init and with the dummy region, in both cases it runs fine the first time and then either seg faults or hangs. When it completes the first time it still has the distorted output as I posted above.

We have the newer cray compilers with clang so I will try with those as well sometime this week and provide an update.

**Comment by daboehme** [comment]
The `create_attribute()` does indeed initialize Caliper, so that shouldn't be the problem per se. There are indeed some issues with initializing Caliper in a multi-threaded region, but the `create_attribute()` should take care of that in this case. Creating a dummy region outside the parallel region first as @jrmadsen suggested is worth a try. There still seems to be some problem with the TLS though, given that we also have the crash in the TLS object destructor.
